### ISSUE TITLE
Can we avoid the dependency on aqt?

### ISSUE THREAD

**Issue by ckp95** [lervag/apy#54]
One of my biggest peeves with `apy` is its implicit dependence on Qt. Basically, the code that lets you sync to Ankiweb relies on importing the `ProfileManager` class, which comes from `aqt`. And `aqt` needs `PyQT` to run. That's a huge pain in the ass to get working on some systems. For example, I just tried installing it again and got

```
Python 3.10.10 (main, Mar  5 2023, 22:26:53) [GCC 12.2.1 20230201] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import aqt
Running with temporary Qt5 compatibility shims.
Run with DISABLE_QT5_COMPAT=1 to confirm compatibility with Qt6.
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/tmp/apy/.venv/lib/python3.10/site-packages/aqt/__init__.py", line 46, in <module>
    from aqt import gui_hooks
  File "/tmp/apy/.venv/lib/python3.10/site-packages/aqt/gui_hooks.py", line 11, in <module>
    from _aqt.hooks import *
  File "/tmp/apy/.venv/lib/python3.10/site-packages/_aqt/hooks.py", line 18, in <module>
    from aqt.qt import QDialog, QEvent, QMenu, QModelIndex, QWidget, QMimeData
  File "/tmp/apy/.venv/lib/python3.10/site-packages/aqt/qt/__init__.py", line 20, in <module>
    from . import qt5_compat  # needs to be imported first
  File "/tmp/apy/.venv/lib/python3.10/site-packages/aqt/qt/qt5_compat.py", line 21, in <module>
    import PyQt6.QtWebEngineCore
ModuleNotFoundError: No module named 'PyQt6.QtWebEngineCore'
```

In my experience there's always some crap like this where a bunch of dependencies for PyQT don't get installed right, and I waste hours debugging it, then forget what I did when it happens the next time months later.

And IMO it's silly for a purely CLI tool to need a GUI framework as a dependency.

I looked a little closer at the code, and unless I'm mistaken, all we need the `ProfileManager` class for is to get a `SyncAuth` object constructed from the profile data. That [only relies](https://github.com/ankitects/anki/blob/main/qt/aqt/profiles.py#L646) on the `syncKey`, `currentSyncUrl`, `customSyncUrl`, and `networkTimeout` keys of the profile. The profile is stored as a pickle blob inside a sqlite file. So couldn't `apy` just pull out all this data itself? That way we could sever the dependency on `aqt`, and only rely on the `anki` package.

**Comment by ckp95** [comment]
Proof of concept:

```python
import sqlite3
import pickle
from pathlib import Path
from contextlib import contextmanager
from pprint import pprint

@contextmanager
def dbconnect(*args, **kwargs):
    conn = sqlite3.connect(*args, **kwargs)
    try:
        yield conn
    finally:
        conn.close()


dbpath = Path.home() / ".local/share/Anki2/prefs21.db"
query = "select name, cast(data as blob) from profiles"

with dbconnect(dbpath) as conn:
    res = conn.execute(query)
    profiles = {name: pickle.loads(data) for name, data in res if name != "_global"}
    for name, data in profiles.items():
        print(name)
        human_readable_data = {k: v for k, v in data.items() if not isinstance(v, bytes)}
        pprint(human_readable_data)
```

Gives me:

```
User 1                                                                                                                                                         
{'allowHTML': False,                                                                                                                                           
 'autoSync': True,                                                                                                                                             
 'autoSyncMediaMinutes': 15,                                                                                                                                   
 'deleteMedia': False,                                                                                                                                         
 'importDirectory': '/home/cameron/me/docs/anki',                                                                                                              
 'importMode': 1,                                                                                                                                              
 'lastColour': '#00f',                                                                                                                                         
 'lastHighlightColor': '#00f',                                                                                                                                 
 'lastOptimize': 1684337679,                                                                                                                                   
 'lastTextColor': '#00f',                                                                                                                                      
 'numBackups': 50,                                                                                                                                             
 'searchHistory': ['abogado'],                                                                                                                                 
 'stripHTML': True,                                                                                                                                            
 'syncKey': None,                                                                                                                                              
 'syncMedia': True} 
```

So it can clearly be opened using stdlib tooling, and it has the `syncKey` entry. (Mine is `None` because I haven't synced to Ankiweb on this computer yet).

**Comment by lervag** [comment]
> One of my biggest peeves with `apy` is its implicit dependence on Qt. …

Yes, I very much agree. For me, things have "just worked", so I've never needed to dig deeper. I managed to get the syncing to work by using `aqt` and did not really dig deeper. But I was also annoyed at having to depend on QT here.

> In my experience there's always some crap like this where a bunch of dependencies for PyQT don't get installed right, and I waste hours debugging it, then forget what I did when it happens the next time months later.

I'm sorry to hear that. I would not mind helping and getting help to resolve this!

> And IMO it's silly for a purely CLI tool to need a GUI framework as a dependency.

Definitely agree!

> I looked a little closer at the code, and unless I'm mistaken, all we need the `ProfileManager` class for is to get a `SyncAuth` object constructed from the profile data.

Yes; but also the `collectionPath`.

> That [only relies](https://github.com/ankitects/anki/blob/main/qt/aqt/profiles.py#L646) on the `syncKey`, `currentSyncUrl`, `customSyncUrl`, and `networkTimeout` keys of the profile. The profile is stored as a pickle blob inside a sqlite file. So couldn't `apy` just pull out all this data itself? That way we could sever the dependency on `aqt`, and only rely on the `anki` package.

Seems like a very good idea to me!

**Comment by lervag** [comment]
> Proof of concept: …

There is very little I like better than the combination of 1) good ideas, and 2) contributions of actual solutions/implementations of good ideas! <3

The current relevant code in `apy` is this:

https://github.com/lervag/apy/blob/f6dc8afea214e5a42c1036e677be588f3693d431/apy/anki.py#L53-L63

What we want is something like what you propose, where `base` is already provided and is the base path (e.g. `~/.local/share/Anki2`). We can unpickle the data and fetch the `collectionPath` and the relevant sync data.

Perhaps we need to add some fallback code or do some extra exception handling, but it should not be to much and too hard!

It seems like a relatively simple update, to be honest. I will definitely do it, but I'm quite busy and won't have the time for perhaps a week or so. However, if you should want to propose a PR yourself, then I will be very happy to review and will prioritize doing it as fast as I can squeeze it in!

**Comment by lervag** [comment]
I believe I've fixed this with #57.
